picard/coverart/providers/__init__.py

```diff
@@ -62,37 +63,52 @@ def __init__(self, parent=None):
 def register_cover_art_provider(provider):
     _cover_art_providers.register(provider.__module__, provider)
     if hasattr(provider, 'OPTIONS') and provider.OPTIONS:
-        provider.OPTIONS.NAME = provider.NAME
-        provider.OPTIONS.TITLE = provider.TITLE or provider.NAME
+        provider.OPTIONS.NAME = provider.name
+        provider.OPTIONS.TITLE = provider.title
         register_options_page(provider.OPTIONS)
 
 
+# named tuples used by cover_art_providers()
+ProviderTuple = namedtuple('ProviderTuple', 'name title enabled cls')
+PInfoTuple = namedtuple('PInfoTuple', 'position enabled')
+POrderTuple = namedtuple('OrderTuple', 'name position enabled')
+
+
 def cover_art_providers():
-    order = [p[0] for p in config.setting['ca_providers']]
+    def from_ca_providers_option():
+        """Iterate through ca_providers option and yield name, position and enabled"""
+        for pos, (name, enabled) in enumerate(config.setting['ca_providers']):
+            yield POrderTuple(name=name, position=pos, enabled=enabled)
 
-    def _key_provider(p):
-        try:
-            return order.index(p.NAME)
-        except ValueError:
-            return 666  # move to the end
-    providers = []
-    for p in sorted(_cover_art_providers, key=_key_provider):
-        providers.append(p)
-    log.debug("CA Providers order: %s",
-              ' > '.join([p.NAME for p in providers]))
-    return providers
-
-
-def is_provider_enabled(provider_name):
-    """Test if provider with name `provider_name` was enabled
-    by user through options"""
-    for name, checked in config.setting['ca_providers']:
-        if name == provider_name:
-            return checked
-    return False
-
-
-class CoverArtProvider(object):
+    # build a defaultdict with provider name as key, and PInfoTuple as value
+    order = defaultdict(lambda: PInfoTuple(position=666, enabled=False))
+    for o in from_ca_providers_option():
+        order[o.name] = PInfoTuple(position=o.position, enabled=o.enabled)
+
+    # use previously built dict to order providers, according to current ca_providers
+    # (yet) unknown providers are placed at the end, disabled
+    ordered_providers = sorted(_cover_art_providers, key=lambda p: order[p.name].position)
+
+    log.debug("CA Providers order: %s", ' > '.join([p.name for p in ordered_providers]))
+
+    for p in ordered_providers:
+        yield ProviderTuple(name=p.name, title=p.title, enabled=order[p.name].enabled, cls=p)
+
+
+class CoverArtProviderMetaClass(type):
```

Comment:
I don't see the use of a meta class here, why not just add those two properties to the `CoverArtProvider` base class?

Reply:
Using `CoverArtProvider` properties didn't work for me,  i tried but they return `property object` instead of `str`.

I made a little [test script](https://gist.github.com/zas/3e461215fe963f6f7e6b33eca67cc124) to show the behavior.
Perhaps there's another way to achieve the expected behavior but i found using a metaclass was doing what i expected.
Ideas / improvements are still welcome.